src/common/src/RnNoiseCommonPlugin.cpp

```diff
@@ -6,84 +6,180 @@
+  m_outBuffer.resize(k_denoiseFrameSize * 2);
+  m_outBufferR = 0;
+  m_outBufferW = 0;
+  m_outBufferA = 0;
 
-        m_outputBuffer.resize(m_outputBuffer.size() + framesToProcess);
+  m_initialized = true;
+}
 
-        // Process input buffer by chunks of k_denoiseFrameSize, put result into out buffer to return into range [-1.f,1.f]
-        {
-            float *outBufferWriteStart = (m_outputBuffer.end() - framesToProcess).base();
+void RnNoiseCommonPlugin::deinit() {
+  m_denoiseState.reset();
+  m_srcDown     .reset();
+  m_srcUp       .reset();
+  m_initialized = false;
+}
 
-            for (size_t i = 0; i < samplesToProcess; i++) {
-                float *currentOutBuffer = &outBufferWriteStart[i * k_denoiseFrameSize];
-                float *currentInBuffer = &m_inputBuffer[i * k_denoiseFrameSize];
-                rnnoise_process_frame(m_denoiseState.get(), currentOutBuffer, currentInBuffer);
+void RnNoiseCommonPlugin::resampleOut(SRC_DATA &sUp, long &frames)
+{
+  // resample what we can to the output
+  while(m_outBufferA && sUp.output_frames)
+  {
+    sUp.data_in       = &m_outBuffer[m_outBufferR];
+    sUp.input_frames  = m_outBufferW < m_outBufferR ? m_outBuffer.size() - m_outBufferR : m_outBufferW - m_outBufferR;
+
+    if (m_resample)
+      src_process(m_srcUp.get(), &sUp);
+    else
+    {
+      // simply copy the buffer if we are not resampling
+      sUp.input_frames_used = sUp.input_frames;
+      if (sUp.input_frames_used > sUp.output_frames)
+        sUp.input_frames_used = sUp.output_frames;
+      memcpy(sUp.data_out, sUp.data_in, sUp.input_frames_used * sizeof(float));
+    }
 
-                for (size_t j = 0; j < k_denoiseFrameSize; j++) {
-                    currentOutBuffer[j] /= std::numeric_limits<short>::max();
-                }
-            }
-        }
+    if (!sUp.input_frames_used && !sUp.output_frames_gen)
+      break;
 
-        const size_t toCopyIntoOutput = std::min(m_outputBuffer.size(), static_cast<size_t>(sampleFrames));
+    m_outBufferR      += sUp.input_frames_used;
+    m_outBufferA      -= sUp.input_frames_used;
 
-        std::memcpy(out, &m_outputBuffer[0], toCopyIntoOutput * sizeof(float));
+    sUp.data_out      += sUp.output_frames_gen;
+    sUp.output_frames -= sUp.output_frames_gen;
+    frames            += sUp.output_frames_gen;
 
-        m_inputBuffer.erase(m_inputBuffer.begin(), m_inputBuffer.begin() + framesToProcess);
-        m_outputBuffer.erase(m_outputBuffer.begin(), m_outputBuffer.begin() + toCopyIntoOutput);
+    if (m_outBufferR == m_outBuffer.size())
+      m_outBufferR = 0;
+  }
+}
 
-        if (toCopyIntoOutput < sampleFrames) {
-            std::fill(out + toCopyIntoOutput, out + sampleFrames, 0.f);
-        }
+void RnNoiseCommonPlugin::process(const float *in, float *out, int32_t sampleFrames)
+{
+  const float mul = 1.0f / std::numeric_limits<short>::max();
+  if (!sampleFrames)
+    return;
+
+  if (!m_initialized)
+    init();
+
+  SRC_DATA sDn;
```

Comment:
Maybe make name more descriptive? Same goes for `sUp`

Reply:
Done :+1: